### PR TITLE
add bump version

### PR DESCRIPTION
**Pull request by slemasne**
# Pull Request Checklist

- [ ] Read the [contributing guidelines](https://github.com/finbourne/lusid-python-tools/blob/master/docs/CONTRIBUTING.md)
- [ ] Tests pass

# Description of the PR

This feature will bump the patch version of **lusid-python-tools** on a deployment to **master** branch.

We use the bump2version package for controlling the versioning. 

https://github.com/c4urself/bump2version

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -12,4 +12,5 @@ coloredlogs==10.*
 detect_delimiter==0.1.*
 pyyaml==5.*
 black==19.10.b0
-flatten-json==0.1.*
\ No newline at end of file
+flatten-json==0.1.*
+bump2version
```
Comment (MikeMcGarry) [inline requirements.txt:16]: Can we please add bump2version and while we are at it also black to a requirements.dev.txt file. They are not required to use LUSID Python Tools only if you are developing.

File: ci_tools/bumpversion.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+bump2version patch --message "Bump version: {current_version} to {new_version} [ci skip]"
+git config --local user.email "engineering@finbourne.com"
+git config --local user.name "finbourne-bot-public"
+git push https://slemasne:${GH_TOKEN}@github.com/slemasne/repo-test.git HEAD:$TRAVIS_BRANCH
```
Comment (MikeMcGarry) [inline ci_tools/bumpversion.sh:4]: What is the reference to the repo-test here for?

File: .travis.yml [hunk 3]
```diff
@@ -39,9 +39,12 @@ jobs:
 
 deploy:
   - provider: script
-    script: bash ci_tools/publish.sh "$fbn_pypi_username" "$fbn_pypi_password"
+    script: >-
+            bash ci_tools/bumpversion.sh &&
+            bash ci_tools/publish.sh "$fbn_pypi_username" "$fbn_pypi_password"
```
Comment (MikeMcGarry) [inline .travis.yml:44]: Where is ci_tools/publish.sh? I don't see it here in this merge request?

File: ci_tools/bumpversion.sh (hunk 2 above)
Comment (rizwansaeed) [reply ci_tools/bumpversion.sh:4]: You also have the wrong username, looks like you’ve left in the test details

File: requirements.txt (hunk 1 above)
Comment (rizwansaeed) [reply requirements.txt:16]: If you crate the new file, you can inherit the main requirements.txt by adding a line with “-r requirements.txt”

File: .travis.yml [hunk 4]
```diff
@@ -39,9 +39,12 @@ jobs:
...
     on: master
 
+
 notifications:
   slack:
     on_success: never
```
Comment (rizwansaeed) [inline .travis.yml:50]: It might be worth posting a slack message when we publish a new version

File: requirements.txt (hunk 1 above)
Comment (slemasne) [reply requirements.txt]: Good idea. That's done in latest commit.

File: ci_tools/bumpversion.sh (hunk 2 above)
Comment (slemasne) [reply ci_tools/bumpversion.sh]: Good catch. Updated in latest commit.

File: .travis.yml (hunk 3 above)
Comment (slemasne) [reply .travis.yml:44]: We added **ci_tools** directory in PR 108.

https://github.com/finbourne/lusid-python-tools/pull/108

This directory contains the continuous integration shell scripts used by the Travis YAML.

File: .bumpversion.cfg [hunk 5]
```diff
@@ -0,0 +1,5 @@
+[bumpversion]
+current_version = 0.2.0
```
[the remark this answers is not in the document]
Comment (slemasne) [reply .bumpversion.cfg:2]: It is dynamic not static. 
It gets incremented when you run:
$ bump2version <part>
The **current_version** here and the version in **__version__.py** need to be in-sync for the package to work.

File: .travis.yml (hunk 4 above)
Comment (slemasne) [reply .travis.yml]: Sure.
Into which chat room? I see the YAML references an encoded room.

File: .bumpversion.cfg (hunk 5 above)
[the remark this answers is not in the document]
Comment (rizwansaeed) [reply .bumpversion.cfg:2]: presumably this gets updated and committed when you bump the version?

File: .travis.yml (hunk 4 above)
Comment (rizwansaeed) [reply .travis.yml]: #build_public is probably ok for now

File: .bumpversion.cfg (hunk 5 above)
[the remark this answers is not in the document]
Comment (slemasne) [reply .bumpversion.cfg:2]: Yes, exactly. That happens on deployment in the **ci_tools/bumpversion.sh** script.

File: .travis.yml (hunk 4 above)
Comment (slemasne) [reply .travis.yml]: Sure, updated.

File: .travis.yml [hunk 6]
```diff
@@ -39,14 +39,28 @@ jobs:
 
 deploy:
   - provider: script
-    script: bash ci_tools/publish.sh "$fbn_pypi_username" "$fbn_pypi_password"
+    script: >-
```
Comment (MikeMcGarry) [inline .travis.yml:42]: Earlier in the Travis file you need to replace
- pip install -r requirements.txt
with
- pip install -r requirements.dev.txt

Comment (MikeMcGarry) [inline .travis.yml:42]: Also earlier in the Travis file I see:

- stage: Run black formatter and commit any changes
      if: branch != master
      script:
        - bash ci_tools/new_commit_with_black.sh

If this triggers, what happens to the rest of the Travis job? Does it test the now out of date code or does it stop given that it has pushed a commit and invalidated the code it is running against?

File: ci_tools/bumpversion.sh [hunk 7]
```diff
@@ -0,0 +1,5 @@
+bump2version patch --message "Bump version: {current_version} to {new_version} [ci skip]"
```
Comment (MikeMcGarry) [inline ci_tools/bumpversion.sh:1]: What does  "[ci skip]" do?

File: ci_tools/bumpversion.sh [hunk 8]
```diff
@@ -0,0 +1,5 @@
...
+git config --local user.email "engineering@finbourne.com"
+git config --local user.name "finbourne-bot-public"
+git push https://finbourne-bot-public:${GH_TOKEN}@github.com/finbourne/lusid-python-tools.git HEAD:$TRAVIS_BRANCH
```
Comment (MikeMcGarry) [inline ci_tools/bumpversion.sh:4]: What does "HEAD" do here? Is it required?

File: .travis.yml (hunk 6 above)
Comment (slemasne) [reply .travis.yml:42]: Ah, of course. I missed that. Updated.

Comment (slemasne) [reply .travis.yml:42]: A new commit triggers a new build which cancels the old build.
This is configured by the auto-cancellation functionality of Travis:
https://blog.travis-ci.com/2017-03-22-introducing-auto-cancellation

File: ci_tools/bumpversion.sh (hunk 7 above)
Comment (slemasne) [reply ci_tools/bumpversion.sh:1]: This skips travis on the commit which updates the version on master.
Without this tag, Travis gets stuck in an infinite loop of:
(1) Update version
(2) This triggers a build
(3) Which updates version
(4) Which triggers another build
(5) And so on...

https://docs.travis-ci.com/user/customizing-the-build/#skipping-a-build

File: ci_tools/bumpversion.sh (hunk 8 above)
Comment (slemasne) [reply ci_tools/bumpversion.sh:4]: Making a commit directly to the branch in Travis causes the HEAD to detach. See logs below. Therefore it seems we need to push this commit back to HEAD again. 


Logs from Travis
----------------------------------------------------------
[0Ktravis_time:start:1c1aaaf4
[0K$ bash ci_tools/new_commit_with_black.sh
[1mreformatted /home/travis/build/slemasne/repo-test/black_test.py[0m
[1mAll done! âœ¨ ðŸ° âœ¨[0m
[1m1 file reformatted[0m, 6 files left unchanged.[0m
**[detached HEAD 7cddf67] black code formatting**
 1 file changed, 1 insertion(+), 4 deletions(-)
Everything up-to-date

**Pull request merged**